app/models/smart_proxy.rb

```diff
@@ -82,18 +82,34 @@ def refresh
   end
 
   def taxonomy_foreign_conditions
-    feature_names = self.features.pluck(:name)
-    conditions                      = {}
-    if feature_names.include?('Puppet') && feature_names.include?('Puppet CA')
+    conditions = {}
+    if has_puppet? && has_puppet_ca?
       conditions = "puppet_proxy_id = #{id} OR puppet_ca_proxy_id = #{id}"
-    elsif feature_names.include?('Puppet')
+    elsif has_puppet?
       conditions[:puppet_proxy_id] = id
-    elsif feature_names.include?('Puppet CA')
+    elsif has_puppet_ca?
       conditions[:puppet_ca_proxy_id] = id
     end
     conditions
   end
 
+  # Helper methods for features - e.g. has_puppet? has_dns?
+  def method_missing(method, *args, &block)
```

Comment:
> Rails tends to use define_method as far as I know, but that then introduces the class-load issue I mentioned in the ticket.

Rails dynamic finders, for example, are method_missing.  That said, they are deprecated in 4 and gone in Rails 5 because the new hash syntax looks similar:

```ruby
Book.find_by title: 'Soylent Green'
```

>  It also avoids needing to update method prototypes for core Ruby methods if they change (#respond_to? has before).

That basically means "don't ever use metaprogramming"

`proxy.has_feature?('DHCP')` does not look as nice (to me) as  `proxy.has_dhcp?`

Anyway, I relent, less magic...